File: atcoder/abc165C_many_requirements.cpp
// Vicfred & unihernandez22
// https://atcoder.jp/contests/abc165/tasks/abc165_c
// dfs, brute force
#include <algorithm>
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

int main() {
    int n, m, q;
    cin >> n >> m >> q;

    vector<int> a(q);
    vector<int> b(q);
    vector<int> c(q);
    vector<int> d(q);

    for(int i = 0; i < q; i++)
        cin >> a[i] >> b[i] >> c[i] >> d[i];

    queue<vector<int>> A;
    for(int i = 1; i <= m; ++i)
        A.push({i});

    while(A.front().size() != n) {
        auto item = A.front();
        A.pop();
        int size = item.size();
        for(int i = item[size-1]; i <= m; i++) {
            item.push_back(i);
            A.push(item);
            item.pop_back();
        }
    }

    long long maxima = 0LL;

    while(!A.empty()) {
        long long local = 0LL;

        auto item = A.front();
        A.pop();

        for(int i = 0; i < q; i++)
            if((item[b[i]-1] - item[a[i]-1]) == c[i])
                local += d[i];

        maxima = max(maxima, local);
    }

    cout << maxima << endl;

    return 0;
}

File: atcoder/abc165C_many_requirements.py
# uninhm
# https://atcoder.jp/contests/abc165/tasks/abc165_c
# dfs, brute force
n, m, q = map(int, input().split())

a = []
b = []
c = []
d = []

for _ in range(q):
    ins = list(map(int, input().split()))
    a.append(ins[0])
    b.append(ins[1])
    c.append(ins[2])
    d.append(ins[3])
 
A = [[i] for i in range(1, m+1)]
 
while len(A[0]) != n:
    current = A.pop(0)
    for i in range(current[-1], m+1):
        A.append(current+[i])
 
maxima = 0
for item in A:
    accumulate = 0
    for i in range(q):
        if (item[b[i]-1] - item[a[i]-1]) == c[i]:
            accumulate += d[i]
    maxima = max(maxima, accumulate)

print(maxima)
